### DISCUSSION TITLE
Pest pretty print for Rules

### DISCUSSION THREAD

**Discussion by Alfus** [pest-parser/pest#823] [Q&A] [answered]
I've been using pest and really find the output on the https://pest.rs/#editor useful!
How do I produce the same output locally?

All the variations of `print!` I tried produced less human readable output. Is there a function somewhere I can call to produce the output from the web site?

Thanks!

**Comment by tomtau** [comment, marked as answer]
Do you mean the pairs formatting? You can check out this function in the website's source code: https://github.com/pest-parser/site/blob/master/src/lib.rs#L114

**Reply by Alfus** [reply]
Yes, thanks! I now have nice, human readable, 'golden' file test output for my grammar.
![image](https://user-images.githubusercontent.com/4567082/224503607-7dfc06a9-2ab1-4dba-90f4-b87de9ab6725.png)
(previously I was outputting something similar, but using serde_yaml)

Maybe consider a feature request to have this format be the normal (or debug) output format for normal Pairs? (I believe this code requires you use the vm, to get Pair<&str>)

**Reply by ndevenish** [reply]
FWIW I copy-pasted the code and the changes to get it to accept the plain `Pair<Rule>` were minimal:
```diff
-fn format_pair(pair: Pair<&str>, indent_level: usize, is_newline: bool) -> String {
+fn format_pair(pair: Pair<Rule>, indent_level: usize, is_newline: bool) -> String {
     let indent = if is_newline {
         "  ".repeat(indent_level)
     } else {
@@ -83,7 +92,7 @@ fn format_pair(pair: Pair<&str>, indent_level: usize, is_newline: bool) -> Strin
     };
     match len {
         0 => format!(
-            "{}{}{}{}: {:?}",
+            "{}{}{}{:?}: {:?}",
             indent,
             dash,
             pair_tag,
@@ -91,7 +100,7 @@ fn format_pair(pair: Pair<&str>, indent_level: usize, is_newline: bool) -> Strin
             pair.as_span().as_str()
         ),
         1 => format!(
-            "{}{}{}{} > {}",
+            "{}{}{}{:?} > {}",
             indent,
             dash,
             pair_tag,
@@ -99,7 +108,7 @@ fn format_pair(pair: Pair<&str>, indent_level: usize, is_newline: bool) -> Strin
             children[0]
         ),
         _ => format!(
-            "{}{}{}{}\n{}",
+            "{}{}{}{:?}\n{}",
             indent,
             dash,
             pair_tag,
```
